File: 733-flood-fill/flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        original_color = image[sr][sc]
        if original_color == color:
            return image
            
        m = len(image)
        n = len(image[0])
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        queue = deque([(sr, sc)])

        while queue:
            r, c = queue.popleft()

            if image[r][c] == original_color:
                image[r][c] = color

                for dr, dc in directions:
                    nr = r + dr
                    nc = c + dc

                    if 0 <= nr < m and 0 <= nc < n and image[nr][nc] == original_color:
                        queue.append((nr, nc))
            
        return image
            
        
        

File: 733-flood-fill/flood-fill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int original_color = image[sr][sc];

        if (image[sr][sc] == color) return image;

        fill(image, sr, sc, color, original_color);

        return image;

    }

    private void fill(int[][] image, int sr, int sc, int color, int original_color) {
        int m = image.length;
        int n = image[0].length;

        if (sr < 0 || sr >= m || sc < 0 || sc >= n || image[sr][sc] != original_color) {
            return;
        }

        image[sr][sc] = color;

        fill(image, sr - 1, sc, color, original_color);
        fill(image, sr + 1, sc, color, original_color);
        fill(image, sr, sc - 1, color, original_color);
        fill(image, sr, sc + 1, color, original_color);


    }

}